aws/resource_aws_launch_template.go

```diff
@@ -0,0 +1,976 @@
+		return err
+	}
+
+	log.Printf("[DEBUG] Received launch template version %q (version %d)", d.Id(), *lt.LatestVersionNumber)
+
+	ltData := dltv.LaunchTemplateVersions[0].LaunchTemplateData
+
+	d.Set("disable_api_termination", ltData.DisableApiTermination)
+	d.Set("ebs_optimized", ltData.EbsOptimized)
+	d.Set("image_id", ltData.ImageId)
+	d.Set("instance_initiated_shutdown_behavior", ltData.InstanceInitiatedShutdownBehavior)
+	d.Set("instance_type", ltData.InstanceType)
+	d.Set("kernel_id", ltData.KernelId)
+	d.Set("key_name", ltData.KeyName)
+	d.Set("monitoring", ltData.Monitoring)
+	d.Set("ram_dist_id", ltData.RamDiskId)
+	d.Set("user_data", ltData.UserData)
+
+	if err := d.Set("block_device_mappings", getBlockDeviceMappings(ltData.BlockDeviceMappings)); err != nil {
+		return err
+	}
+
+	if err := d.Set("credit_specification", getCreditSpecification(ltData.CreditSpecification)); err != nil {
+		return err
+	}
+
+	if err := d.Set("elastic_gpu_specifications", getElasticGpuSpecifications(ltData.ElasticGpuSpecifications)); err != nil {
+		return err
+	}
+
+	if err := d.Set("iam_instance_profile", getIamInstanceProfile(ltData.IamInstanceProfile)); err != nil {
+		return err
+	}
+
+	if err := d.Set("instance_market_options", getInstanceMarketOptions(ltData.InstanceMarketOptions)); err != nil {
+		return err
+	}
+
+	if err := d.Set("network_interfaces", getNetworkInterfaces(ltData.NetworkInterfaces)); err != nil {
+		return err
+	}
+
+	if err := d.Set("placement", getPlacement(ltData.Placement)); err != nil {
+		return err
+	}
+
+	if err := d.Set("tag_specifications", getTagSpecifications(ltData.TagSpecifications)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourceAwsLaunchTemplateUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	if !d.IsNewResource() {
+		launchTemplateData, err := buildLaunchTemplateData(d, meta)
+		if err != nil {
+			return err
+		}
+
+		launchTemplateVersionOpts := &ec2.CreateLaunchTemplateVersionInput{
+			ClientToken:        aws.String(resource.UniqueId()),
+			LaunchTemplateId:   aws.String(d.Id()),
+			LaunchTemplateData: launchTemplateData,
+		}
+
+		_, createErr := conn.CreateLaunchTemplateVersion(launchTemplateVersionOpts)
+		if createErr != nil {
+			return createErr
+		}
+	}
+
+	return resourceAwsLaunchTemplateRead(d, meta)
+}
+
+func resourceAwsLaunchTemplateDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	log.Printf("[DEBUG] Launch Template destroy: %v", d.Id())
+	_, err := conn.DeleteLaunchTemplate(&ec2.DeleteLaunchTemplateInput{
+		LaunchTemplateId: aws.String(d.Id()),
+	})
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Launch Template deleted: %v", d.Id())
+	return nil
+}
+
+func getBlockDeviceMappings(m []*ec2.LaunchTemplateBlockDeviceMapping) []interface{} {
+	s := []interface{}{}
+	for _, v := range m {
+		mapping := map[string]interface{}{
+			"device_name":  *v.DeviceName,
```

Comment:
Rather than risk panics on nil dereferencing, we should switch anything using `*thing` to the appropriate SDK method that will set it to the zero value for the type: e.g. `aws.StringValue(v.DeviceName)` This should repeated everywhere below.